### PR TITLE
proper ckpt recovery

### PR DESCRIPTION
**Pull request by samsja**
Screenshot from 2024-10-16 00-12-24

### CODE REVIEW

File: src/zeroband/utils/state_dict_send_recv.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            if isinstance(value, dict):
+                new_dict[key] = _split(value, tensors_)
+            elif isinstance(value, torch.Tensor):
+                idx = len(tensors_)
+                tensors_.append(value)
+                new_dict[key] = _tensor_to_placeholder(idx, value)
+            else:
+                new_dict[key] = value
+
+        return new_dict
```
Comment (Jackmin801) [inline src/zeroband/utils/state_dict_send_recv.py]: Just a nit. Im not sure if the list is necessary here. It seems to me that the logic only ever sets children into the new_dict. Given that the state dict is a tree, the current `state_dict_` should not see any modifications during the loop.

File: src/zeroband/utils/state_dict_send_recv.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+def _get_sendable_state_dict(state_dict: dict) -> tuple[dict, list[torch.Tensor]]:
+    """
+    This function take a state dict (dict with tensor inside) and return a torch.send/recv-able format.
+
+    It splits the state dict into two part :
+    * a list of tensor
+    * a dict emptied from tensor
+
+    The order is deterministic. The function can be used in pair with  load_sendable_opt_state
```
Comment (Jackmin801) [inline src/zeroband/utils/state_dict_send_recv.py]: Does `load_sendable_opt_state` still exist?

File: src/zeroband/utils/state_dict_send_recv.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    state_dict = _split(state_dict, tensors)
+    return state_dict, tensors
+
+
+def _load_sendable_state_dict(tensors: list[torch.Tensor], state_dict: dict) -> dict:
+    """
+    This function take a list of tensor and a state dict and return state dict.
+
+    The function can be used in pair with load_sendable_opt_state
```
Comment (Jackmin801) [inline src/zeroband/utils/state_dict_send_recv.py]: Does `load_sendable_opt_state` still exist?

File: src/zeroband/utils/state_dict_send_recv.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    """
+
+    def _load(state_dict_):
+        for key, value in list(state_dict_.items()):  # list needed as we modify the state_dict_ as we traverse it
+            if isinstance(value, dict):
+                state_dict_[key] = _load(value)
+            if isinstance(value, str) and value.startswith("zeroband_tensor_"):
+                state_dict_[key] = _validate_placeholder_to_tensor(value, tensors)
+        return state_dict_
```
Comment (Jackmin801) [inline src/zeroband/utils/state_dict_send_recv.py]: Is there a reason we only load tensors?

File: src/zeroband/utils/state_dict_send_recv.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    state_dict_tensor_buffer, size = _object_to_tensor(state_dict)
+    pg.send([size], dest_rank, 0).wait()
+    pg.send([state_dict_tensor_buffer], dest_rank, 0).wait()
+
+    for i, tensor in enumerate(tensors):
+        buffer = tensor
+        if isinstance(tensor, DTensor):
+            buffer = tensor.to_local()
+
+        buffer = buffer.cpu()
```
Comment (Jackmin801) [inline src/zeroband/utils/state_dict_send_recv.py]: We might need a `detach()` required here

File: src/zeroband/utils/state_dict_send_recv.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+        pg.send([buffer], dest_rank, 0).wait()
```
Comment (Jackmin801) [inline src/zeroband/utils/state_dict_send_recv.py]: What do you think about setting the tag value to `i`? Also does it make sense to do all the waits at the end? Would allow the sending to be async.

File: src/zeroband/utils/state_dict_send_recv.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    state_dict = _tensor_to_object(object_tensor, size)
+
+    _, tensors = _get_sendable_state_dict(og_state_dict)
+
+    for i, tensor in enumerate(tensors):
+        buffer = tensor
+        if isinstance(tensor, DTensor):
+            buffer = tensor.to_local()
+
+        data = torch.empty_like(buffer).to("cpu")
```
Comment (Jackmin801) [inline src/zeroband/utils/state_dict_send_recv.py]: Could we set the device to "cpu" in the empty_like?

File: src/zeroband/train.py [hunk 8]
```diff
@@ -299,6 +315,8 @@ def train(config: Config):
 
     logger.info("starting training")
 
+    diloco.outer_optimizer.step()  # TODO remove. If you are a reviewer and see its still yet block this PR
```
Comment (Jackmin801) [inline src/zeroband/train.py]: Request granted. Blocked!

**Review by Jackmin801 — CHANGES_REQUESTED** [review]
Past sami asked me to block because a line wasnt removed

File: src/zeroband/utils/state_dict_send_recv.py (hunk 1 above)
Comment (samsja) [reply src/zeroband/utils/state_dict_send_recv.py]: you mean the tensors list ? If yes the idea is that I want to have a flatten list of tensor for the sending/rcv operation

File: src/zeroband/utils/state_dict_send_recv.py (hunk 4 above)
Comment (samsja) [reply src/zeroband/utils/state_dict_send_recv.py:95]: no that's a bug

Comment (samsja) [reply src/zeroband/utils/state_dict_send_recv.py]: this one modify the state dict in place to load the tensors. Hence non tensor value just need to be kept as is

File: src/zeroband/utils/state_dict_send_recv.py (hunk 6 above)
Comment (samsja) [reply src/zeroband/utils/state_dict_send_recv.py]: actually, I don't know fully understand what the tag is doing :sweat: .\

I am planning on doing this async, I want to benchmark with and wo first tho

File: src/zeroband/train.py (hunk 8 above)
Comment (samsja) [reply src/zeroband/train.py]: haha

File: src/zeroband/train.py [hunk 9]
```diff
@@ -256,25 +255,42 @@ def train(config: Config):
...
 
-        if config.train.log_model_hash:
-            logger.debug("Pre diloco model: %s", get_module_signature(model))
+        ## we create grad buffer and opts stats mamnually, the value will be overwritten by the ckpt but we need the DTensor to be correctly init before loading it
 
-        diloco.step(model, fake=True, flag=training_progress.outer_step)
+        diloco.outer_optimizer.step()  # need to step to init the DTensor stats
+
+        ## do a forward backward to init the DTensor stats
+        input_ids = torch.zeros(config.train.micro_bs, config.data.seq_length).long().to("cuda")
```
Comment (Jackmin801) [inline src/zeroband/train.py]: Can we do the long and device in the torch.zeros call?

File: src/zeroband/train.py [hunk 10]
```diff
@@ -256,25 +255,42 @@ def train(config: Config):
...
+        loss = model(tokens=input_ids).contiguous().mean()
+        loss.backward()
+        inner_optimizer.step()
+        inner_optimizer.zero_grad()
```
Comment (Jackmin801) [inline src/zeroband/train.py]: Theres a slightly cheaper way to do this by doing
```
for group in inner_optimizer.param_groups:
    for p in group['params']:
        p.grad = torch.randn_like(p)
```
I believe this results in no all_gathers at all

File: src/zeroband/train.py [hunk 11]
```diff
@@ -318,6 +336,24 @@ def train(config: Config):
...
+                    logger.info(f"inner optimizer hash: {get_optimizer_signature(inner_optimizer)}")
+
+                try:
+                    ckpt_manager.send_ckpt_to_peer(elastic_device_mesh.global_pg, maybe_dest_rank)
+                except RuntimeError as e:
+                    # error can happen if the receiver drop off. We don't want to stop the training if this happend.
+                    logger.error(f"Error sending ckpt to rank {maybe_dest_rank}: {e}")
+
+                elastic_device_mesh.live_recovery.reset()
+
```
Comment (Jackmin801) [inline src/zeroband/train.py]: Eh is this blocking?

File: src/zeroband/train.py [hunk 12]
```diff
@@ -511,7 +547,6 @@ def train(config: Config):
     # However, in development, we want to know that we broke torch compile
     torch._dynamo.config.suppress_errors = "ZERO_BAND_DEV" not in os.environ
     torch.set_float32_matmul_precision("high")
-    torch.manual_seed(42)  # this ensure same weight init across diloco workers
```
Comment (Jackmin801) [inline src/zeroband/train.py:514]: Oh is it not necessary to have this anymore?

File: src/zeroband/train.py (hunk 10 above)
Comment (samsja) [reply src/zeroband/train.py]: the reason I did not go with this solution is that I did not want to mess with the DTensor creation.

 Would torch.zero_like(p) with p being a dtensor do the job ?

I think that I am being paranoid because last time we did this was across device so the DTensor were messed up. Here its not the case

File: src/zeroband/train.py (hunk 12 above)
Comment (samsja) [reply src/zeroband/train.py:514]: no its necessary, bad manip on my side

File: src/zeroband/utils/state_dict_send_recv.py (hunk 1 above)
Comment (Jackmin801) [reply src/zeroband/utils/state_dict_send_recv.py]: Oh I mean the loop
`for key, value in list(state_dict_.items())` should be replaceable by `for key, value in state_dict_.items()`

It shouldnt be too expensive so might be fine to keep as is and not have to test it. But maybe the comment can be removed so a future person isnt confused

File: src/zeroband/train.py (hunk 11 above)
Comment (samsja) [reply src/zeroband/train.py]: nop the send_ckpt_to_peer call a thread which is only awaited before the outer step

Comment (samsja) [reply src/zeroband/train.py]: so the try catch is probably useless right ? I should move it to inside the thread

File: src/zeroband/utils/state_dict_send_recv.py (hunk 6 above)
Comment (Jackmin801) [reply src/zeroband/utils/state_dict_send_recv.py]: Oh the tag is used to match up tensors being send and recv if there are multiple in-flight. Say sender sends 3 tensors at the same time, and tags them 0, 1, 2. It doesnt have a guarantee on arrival order, but if I receiver tags its recv, gloo knows which tensors to set.

File: src/zeroband/train.py (hunk 10 above)
Comment (Jackmin801) [reply src/zeroband/train.py]: I suppose we can test it. But in this particular case I dont see why zeros_like wont work. You can leave it for now if you think its a better idea to not mess with it in this PR

File: src/zeroband/train.py (hunk 11 above)
Comment (Jackmin801) [reply src/zeroband/train.py]: I suppose you can still try catch the async proc not being created correctly? Could be better to move the try catch into send_ckpt_to_peer itself and not increase lines required in train though.

File: src/zeroband/utils/state_dict_send_recv.py (hunk 1 above)
Comment (samsja) [reply src/zeroband/utils/state_dict_send_recv.py]: ah yeah yeah good catch we don't need loop. I used to edit the dict directly

File: src/zeroband/train.py (hunk 10 above)
Comment (samsja) [reply src/zeroband/train.py]: will do then, definitely ugly rn, and it triggers torch compile on potentially a wrong input (ex no seqlens)

File: src/zeroband/utils/state_dict_send_recv.py (hunk 4 above)
Comment (Jackmin801) [reply src/zeroband/utils/state_dict_send_recv.py]: but we send more than just the tensors right? I think every non-tensor is the same except for `lr` for the current run though.

File: src/zeroband/train.py (hunk 10 above)
Comment (samsja) [reply src/zeroband/train.py]: working

**Review by Jackmin801 — APPROVED** [review]
Nice! LFGTM 🚀
